src/main/java/dev/morling/onebrc/CalculateAverage_ericxiao.java

```diff
@@ -137,23 +100,37 @@ public void materialize() {
 
         public void add(long keyStart, long keyEnd, long valueEnd) {
             int entryLength = (int) (valueEnd - keyStart);
+
             int keyLength = (int) (keyEnd - keyStart);
-            int valueLength = (int) (valueEnd - (keyEnd + 1));
             UNSAFE.copyMemory(null, keyStart, entryBytes, Unsafe.ARRAY_BYTE_BASE_OFFSET, entryLength);
-            // String key = new String(entryBytes, 0, keyLength, StandardCharsets.UTF_8);
             KeySlice key = new KeySlice(entryBytes, keyLength);
-            double value = Double.parseDouble(new String(entryBytes, keyLength + 1, valueLength, StandardCharsets.UTF_8));
+
+            int valueLength = (int) (valueEnd - (keyEnd + 1));
+
+            final byte negativeSign = '-';
+            final byte periodSign = '.';
+            int accumulator = 0;
+            int multiplier = 1;
+
+            for (int i = keyLength + 1; i <= keyLength + valueLength; ++i) {
+                if (entryBytes[i] == negativeSign) {
```

Comment:
we can do this outside the loop right ?.

Reply:
Done.